### PR TITLE
Group matcher algorithm

### LINKED ISSUE #386: Group Matcher Algorithm

### PR DESCRIPTION
**Pull request by ketphan02**
Closes #386

Branched out from #331

### CODE REVIEW

File: simulation_cache
Comment (Opeyem1a) [inline simulation_cache:1]: Why does this PR affect the simulation cache at all?

**Review by Opeyem1a — CHANGES_REQUESTED** [review]
Largely trusting you that this works, seems fine though

File: simulation_cache
Comment (ketphan02) [reply simulation_cache:1]: Properly because of auto linting, I'll revert that

Comment (ketphan02) [reply simulation_cache:1]: This is from the `simulation_cache` repo, I run update on the submodules so it automatically update

**Review by Opeyem1a — APPROVED** [review]

**Pull request merged**